### PR TITLE
feat: update to Soldeer v0.4.0

### PR DESCRIPTION
**Pull request by mario-eth**
## Motivation
We finally have the version 0.4.0 ready. 
In 0.4.0 we kinda did a rewrite to have it more rust idiomatic and solved also bugs that were waiting to be fixed.

https://github.com/mario-eth/soldeer/pull/202 
Full description of the release! 

Important mention:
we did split the project into 3 crates:
- cli - used only for `soldeer` alone
- commands - used for commands alone
- core - core crate

We did not publish the crates yet as we are waiting for feedback here beforehand.

### CODE REVIEW

File: crates/forge/tests/cli/soldeer.rs [hunk 1]
```diff
@@ -1,30 +1,41 @@
 //! Contains various tests related to `forge soldeer`.
 
 use std::{
-    fs::{self, OpenOptions},
+    fs::{self},
```
Comment (beeb) [inline crates/forge/tests/cli/soldeer.rs:4]:
```suggestion
    fs,
```

File: crates/forge/tests/cli/soldeer.rs [hunk 2]
```diff
@@ -210,13 +231,9 @@ mario-custom-tag = { version = "1.0", git = "https://gitlab.com/mario4582928/Mar
...
     let foundry_file = prj.root().join("foundry.toml");
+    fs::write(&foundry_file, foundry_updates).unwrap();
 
-    let mut file = OpenOptions::new().append(true).open(&foundry_file).unwrap();
-
-    if let Err(e) = write!(file, "{foundry_updates}") {
-        eprintln!("Couldn't write to file: {e}");
-    }
-
+    cmd.env("NO_COLOR", "");
```
Comment (beeb) [inline crates/forge/tests/cli/soldeer.rs:236]: Not needed in the end

File: crates/forge/tests/cli/soldeer.rs [hunk 3]
```diff
@@ -297,7 +275,7 @@ mario = { version = "1.0", git = "https://gitlab.com/mario4582928/Mario.git", re
 mario-custom-tag = { version = "1.0", git = "https://gitlab.com/mario4582928/Mario.git", tag = "custom-tag" }
 mario-custom-branch = { version = "1.0", git = "https://gitlab.com/mario4582928/Mario.git", tag = "custom-branch" }
 "#;
-
+    println!("str {:?}", read_file_to_string(&foundry_file));
```
Comment (beeb) [inline crates/forge/tests/cli/soldeer.rs:278]: Can probably be removed?

**Review by mattsse — COMMENTED** [review]
this seems fine, using commands directly makes sense

**Review by beeb — APPROVED** [review]
Would love to see this merged before I leave on vacation end of next week!

**Comment by DaniPopes** [comment]
Blocked on https://github.com/mario-eth/soldeer/pull/202

File: crates/forge/tests/cli/soldeer.rs [hunk 4]
```diff
@@ -505,6 +465,21 @@ Added all dependencies to remapppings
     assert_data_eq!(read_file_to_string(&remappings_file), remappings_content);
 });
 
+forgesoldeer!(login, |prj, cmd| {
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/forge/tests/cli/soldeer.rs:468]: sure

File: crates/forge/tests/cli/soldeer.rs [hunk 5]
```diff
@@ -1,30 +1,40 @@
...
+│
+◆  Done reading config
+│  
+[6A[1A
+[2K[1B[1A◇  Installing forge-std~1.8.1
+[6B◆  Dependency added to config
+│  
+◆  Dependency added to lockfile
+│  
+◆  Dependency added to remappings
```
Comment (mattsse) [inline crates/forge/tests/cli/soldeer.rs:35]: looks like we need to escape ◇ here

Comment (beeb) [reply crates/forge/tests/cli/soldeer.rs:35]: Removing the weird characters before the diamonds also works, but yeah probably easier to just wildcard it

**Review by mattsse — APPROVED** [review]

**Pull request merged**